File: windows/src/SecureVox.Core/Models/TranscriptSegment.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace SecureVox.Core.Models;

/// <summary>
/// A segment of transcribed text with timing information
/// </summary>
public class TranscriptSegment
{
    [Key]
    public Guid Id { get; set; } = Guid.NewGuid();

    [Required]
    public string Text { get; set; } = string.Empty;

    /// <summary>
    /// Start time in seconds
    /// </summary>
    public double StartTime { get; set; }

    /// <summary>
    /// End time in seconds
    /// </summary>
    public double EndTime { get; set; }

    /// <summary>
    /// Confidence score (0.0 - 1.0)
    /// </summary>
    public float Confidence { get; set; } = 1.0f;

    /// <summary>
    /// Optional speaker identification
    /// </summary>
    public string? SpeakerLabel { get; set; }

    /// <summary>
    /// Segment index within the recording
    /// </summary>
    public int SegmentIndex { get; set; }

    // Navigation property
    public Guid RecordingId { get; set; }
    public Recording? Recording { get; set; }

    // Computed properties
    [NotMapped]
    public TimeSpan Duration => TimeSpan.FromSeconds(EndTime - StartTime);

    [NotMapped]
    public string FormattedStartTime => FormatTime(StartTime);

    [NotMapped]
    public string FormattedEndTime => FormatTime(EndTime);

    [NotMapped]
    public string FormattedTimeRange => $"{FormattedStartTime} - {FormattedEndTime}";

    /// <summary>
    /// SRT format timestamp for start time (00:00:00,000)
    /// </summary>
    [NotMapped]
    public string SrtStartTimestamp => FormatSrtTime(StartTime);

    /// <summary>
    /// SRT format timestamp for end time (00:00:00,000)
    /// </summary>
    [NotMapped]
    public string SrtEndTimestamp => FormatSrtTime(EndTime);

    /// <summary>
    /// VTT format timestamp for start time (00:00:00.000)
    /// </summary>
    [NotMapped]
    public string VttStartTimestamp => FormatVttTime(StartTime);

    /// <summary>
    /// VTT format timestamp for end time (00:00:00.000)
    /// </summary>
    [NotMapped]
    public string VttEndTimestamp => FormatVttTime(EndTime);

    [NotMapped]
    public bool HasValidTiming => StartTime >= 0 && EndTime >= StartTime;

    private static string FormatTime(double timeInSeconds)
    {
        var time = Math.Max(0, timeInSeconds);
        var minutes = (int)time / 60;
        var seconds = (int)time % 60;
        return $"{minutes}:{seconds:D2}";
    }

    /// <summary>
    /// Format time for SRT subtitle format (00:00:00,000)
    /// </summary>
    private static string FormatSrtTime(double timeInSeconds)
    {
        var time = Math.Max(0, timeInSeconds);
        var hours = (int)time / 3600;
        var minutes = ((int)time % 3600) / 60;
        var seconds = (int)time % 60;
        var milliseconds = (int)((time % 1) * 1000);
        return $"{hours:D2}:{minutes:D2}:{seconds:D2},{milliseconds:D3}";
    }

    /// <summary>
    /// Format time for VTT subtitle format (00:00:00.000)
    /// </summary>
    private static string FormatVttTime(double timeInSeconds)
    {
        var time = Math.Max(0, timeInSeconds);
        var hours = (int)time / 3600;
        var minutes = ((int)time % 3600) / 60;
        var seconds = (int)time % 60;
        var milliseconds = (int)((time % 1) * 1000);
        return $"{hours:D2}:{minutes:D2}:{seconds:D2}.{milliseconds:D3}";
    }
}

File: macos/SecureVox/Models/TranscriptSegment.swift
import Foundation
import SwiftData

/// A segment of transcribed text with timing information
@Model
final class TranscriptSegment {

    // MARK: - Properties

    @Attribute(.unique) var id: UUID
    var text: String
    var startTime: TimeInterval
    var endTime: TimeInterval
    var confidence: Float
    var speakerLabel: String?

    // MARK: - Relationships

    var recording: Recording?

    // MARK: - Computed Properties

    var duration: TimeInterval {
        endTime - startTime
    }

    var formattedStartTime: String {
        formatTime(startTime)
    }

    var formattedEndTime: String {
        formatTime(endTime)
    }

    var formattedTimeRange: String {
        "\(formattedStartTime) - \(formattedEndTime)"
    }

    /// SRT format timestamp for start time (00:00:00,000)
    var srtStartTimestamp: String {
        formatSRTTime(startTime)
    }

    /// SRT format timestamp for end time (00:00:00,000)
    var srtEndTimestamp: String {
        formatSRTTime(endTime)
    }

    /// VTT format timestamp for start time (00:00:00.000)
    var vttStartTimestamp: String {
        formatVTTTime(startTime)
    }

    /// VTT format timestamp for end time (00:00:00.000)
    var vttEndTimestamp: String {
        formatVTTTime(endTime)
    }

    /// Whether this segment has valid timing information
    var hasValidTiming: Bool {
        startTime >= 0 && endTime >= startTime
    }

    // MARK: - Initialization

    init(
        id: UUID = UUID(),
        text: String,
        startTime: TimeInterval,
        endTime: TimeInterval,
        confidence: Float = 1.0,
        speakerLabel: String? = nil,
        recording: Recording? = nil
    ) {
        self.id = id
        self.text = text
        self.startTime = startTime
        self.endTime = endTime
        self.confidence = confidence
        self.speakerLabel = speakerLabel
        self.recording = recording
    }

    // MARK: - Helpers

    private func formatTime(_ time: TimeInterval) -> String {
        let minutes = Int(time) / 60
        let seconds = Int(time) % 60
        return String(format: "%d:%02d", minutes, seconds)
    }

    /// Format time for SRT subtitle format (00:00:00,000)
    private func formatSRTTime(_ time: TimeInterval) -> String {
        let normalizedTime = max(0, time)
        let hours = Int(normalizedTime) / 3600
        let minutes = (Int(normalizedTime) % 3600) / 60
        let seconds = Int(normalizedTime) % 60
        let milliseconds = Int((normalizedTime.truncatingRemainder(dividingBy: 1)) * 1000)
        return String(format: "%02d:%02d:%02d,%03d", hours, minutes, seconds, milliseconds)
    }

    /// Format time for VTT subtitle format (00:00:00.000)
    private func formatVTTTime(_ time: TimeInterval) -> String {
        let normalizedTime = max(0, time)
        let hours = Int(normalizedTime) / 3600
        let minutes = (Int(normalizedTime) % 3600) / 60
        let seconds = Int(normalizedTime) % 60
        let milliseconds = Int((normalizedTime.truncatingRemainder(dividingBy: 1)) * 1000)
        return String(format: "%02d:%02d:%02d.%03d", hours, minutes, seconds, milliseconds)
    }
}
